src/toast/cc-toaster.js

```diff
@@ -0,0 +1,432 @@
+    this.toastDefaultConfig = null;
+
+    /** @type {Array<Toast>} Array of toasts currently displayed. This property should not be used directly. Instead, you should dispatch a `cc-toast:show` event. */
+    this._toasts = [];
+
+    this._dismissingSet = new Set();
+  }
+
+  /* region PRIVATE METHODS */
+  /**
+   * Based on the given toast, forges a new toast where the default config is applied, and where the key is generated if not present.
+   *
+   * @param {Toast} toast
+   * @returns {Toast}
+   * @private
+   */
+  _completeToast (toast) {
+    return {
+      ...toast,
+      key: toast.key ?? Math.random().toString(36).slice(2),
+      config: {
+        ...this.toastDefaultConfig,
+        ...toast.config,
+      },
+    };
+  }
+
+  /**
+   * @param {string} key
+   * @returns {Toast}
+   * @private
+   */
+  _getToastByKey (key) {
+    return this._toasts.find((t) => t.key === key);
+  }
+
+  /**
+   * Shows a new toast.
+   *
+   * If the given toast doesn't contain a `key` property, a random key is generated automatically.
+   * If a toast with the given key already exists, nothing happens.
+   * If the maxToasts is reached, the oldest toast will be dismissed automatically.
+   * If the maxToasts is 0, nothing happens.
+   *
+   * @param {Toast} toast the toast to show
+   * @returns {Toast|undefined} the completed toast. Returns undefined if no toast was shown.
+   */
+  _show (toast) {
+    if (this.maxToasts === 0) {
+      return;
+    }
+
+    // dismiss last toast if max is reached
+    if (this.maxToasts) {
+      const activeToasts = this._toasts.filter((toast) => this._isDismissing(toast.key) !== true);
+
+      if (activeToasts.length >= this.maxToasts) {
+        this._dismiss(activeToasts[activeToasts.length - 1].key);
+      }
+    }
+
+    const completedToast = this._completeToast(toast);
+
+    // do nothing if a toast with the same key already exists.
+    if (this._toasts.find((toast) => toast.key === completedToast.key)) {
+      return;
+    }
+
+    this._toasts = [completedToast, ...this._toasts];
+
+    return completedToast;
+  }
+
+  _dismiss (key) {
+    dispatchCustomEvent(this, 'cc-toast:dismiss', { key, timeoutExpired: false });
+  }
+
+  _startHide (key) {
+    const toast = this._getToastByKey(key);
+    if (toast) {
+      this._dismissingSet.add(toast.key);
+      this.requestUpdate();
+    }
+  }
+
+  /**
+   * @param {AnimationEvent} e
+   * @private
+   */
+  // TODO: in the future (when we have Lit2) we may want to implement that with https://github.com/lit/lit/tree/main/packages/labs/motion
+  _onHideAnimationEnd (e) {
+    const toast = this._getToastByKey(e.target.key);
+    if (toast) {
+      this._toasts = this._toasts.filter((t) => t !== toast);
+      this._dismissingSet.delete(toast.key);
```

Comment:
This is weird because the `this._dismissingSet.delete(toast.key);` requires a `this.requestUpdate();` but the presence of `this._toasts = ` makes it useless...